File: gateOne/daySeven/bankeBank/AccountTest.java
package daySeven.bankeBank;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

public class AccountTest {

    @Test
    public void test_CreateAccount_Success() {
        Account account = new Account("John", "Doe", 1234, "1000000001");
        assertEquals("John", account.getFirstName());
        assertEquals("Doe", account.getLastName());
        assertEquals("1000000001", account.getAccountNumber());
        assertEquals(0.0, account.getBalance());
    }

    @Test
    public void test_Deposit_Success() {
        Account account = new Account("John", "Doe", 1234, "1000000001");
        account.deposit(500.0);
        assertEquals(500.0, account.getBalance());
    }

    @Test
    public void test_Deposit_InvalidAmount() {
        Account account = new Account("John", "Doe", 1234, "1000000001");
        Exception exception = assertThrows(IllegalArgumentException.class, () -> {
            account.deposit(-100.0);
        });
        assertEquals("Deposit amount must be greater than zero", exception.getMessage());
    }

    @Test
    public void test_Withdraw_Success() {
        Account account = new Account("John", "Doe", 1234, "1000000001");
        account.deposit(500.0);
        account.withdraw(200.0, 1234);
        assertEquals(300.0, account.getBalance());
    }

    @Test
    public void test_Withdraw_InvalidAmount() {
        Account account = new Account("John", "Doe", 1234, "1000000001");
        account.deposit(500.0);
        Exception exception = assertThrows(IllegalArgumentException.class, () -> {
            account.withdraw(-100.0, 1234);
        });
        assertEquals("Withdrawal amount must be greater than zero", exception.getMessage());
    }

    @Test
    public void test_Withdraw_InsufficientFunds() {
        Account account = new Account("John", "Doe", 1234, "1000000001");
        account.deposit(100.0);
        Exception exception = assertThrows(IllegalArgumentException.class, () -> {
            account.withdraw(200.0, 1234);
        });
        assertEquals("Insufficient funds", exception.getMessage());
    }

    @Test
    public void test_Withdraw_IncorrectPin() {
        Account account = new Account("John", "Doe", 1234, "1000000001");
        account.deposit(500.0);
        Exception exception = assertThrows(IllegalArgumentException.class, () -> {
            account.withdraw(100.0, 1111);
        });
        assertEquals("Incorrect PIN", exception.getMessage());
    }

    @Test
    public void test_ChangePin_Success() {
        Account account = new Account("John", "Doe", 1234, "1000000001");
        account.changePin(1234, 5678);
        assertTrue(account.verifyPin(5678));
    }

    @Test
    public void test_ChangePin_IncorrectOldPin() {
        Account account = new Account("John", "Doe", 1234, "1000000001");
        Exception exception = assertThrows(IllegalArgumentException.class, () -> {
            account.changePin(1111, 5678);
        });
        assertEquals("Old PIN is incorrect", exception.getMessage());
    }

    @Test
    public void test_ChangePin_InvalidNewPin() {
        Account account = new Account("John", "Doe", 1234, "1000000001");
        Exception exception = assertThrows(IllegalArgumentException.class, () -> {
            account.changePin(1234, 56789);
        });
        assertEquals("New PIN must be exactly 4 digits", exception.getMessage());
    }

    @Test
    public void test_VerifyPin_Success() {
        Account account = new Account("John", "Doe", 1234, "1000000001");
        assertTrue(account.verifyPin(1234));
    }

    @Test
    public void test_VerifyPin_Incorrect() {
        Account account = new Account("John", "Doe", 1234, "1000000001");
        assertFalse(account.verifyPin(1111));
    }
}

File: gateOne/daySeven/bankeBank/Python/AccountTest.py
import unittest
from Account import Account

class AccountTest(unittest.TestCase):
    def test_create_account_success(self):
        account = Account("John", "Doe", 1234, "1000000001")
        self.assertEqual(account.get_first_name(), "John")
        self.assertEqual(account.get_last_name(), "Doe")
        self.assertEqual(account.get_account_number(), "1000000001")
        self.assertEqual(account.get_balance(), 0.0)

    def test_deposit_success(self):
        account = Account("John", "Doe", 1234, "1000000001")
        account.deposit(500.0)
        self.assertEqual(account.get_balance(), 500.0)

    def test_deposit_invalid_amount(self):
        account = Account("John", "Doe", 1234, "1000000001")
        with self.assertRaises(ValueError) as context:
            account.deposit(-100.0)
        self.assertEqual(str(context.exception), "Deposit amount must be greater than zero")

    def test_withdraw_success(self):
        account = Account("John", "Doe", 1234, "1000000001")
        account.deposit(500.0)
        account.withdraw(200.0, 1234)
        self.assertEqual(account.get_balance(), 300.0)

    def test_withdraw_invalid_amount(self):
        account = Account("John", "Doe", 1234, "1000000001")
        account.deposit(500.0)
        with self.assertRaises(ValueError) as context:
            account.withdraw(-100.0, 1234)
        self.assertEqual(str(context.exception), "Withdrawal amount must be greater than zero")

    def test_withdraw_insufficient_funds(self):
        account = Account("John", "Doe", 1234, "1000000001")
        account.deposit(100.0)
        with self.assertRaises(ValueError) as context:
            account.withdraw(200.0, 1234)
        self.assertEqual(str(context.exception), "Insufficient funds")

    def test_withdraw_incorrect_pin(self):
        account = Account("John", "Doe", 1234, "1000000001")
        account.deposit(500.0)
        with self.assertRaises(ValueError) as context:
            account.withdraw(100.0, 1111)
        self.assertEqual(str(context.exception), "Incorrect PIN")

    def test_withdraw_correct_pin(self):
        account = Account("John", "Doe", 1234, "1000000001")
        account.deposit(500.0)
        account.withdraw(100.0, 1234)
        self.assertEqual(account.get_balance(), 400.0)



    def test_change_pin_success(self):
        account = Account("John", "Doe", 1234, "1000000001")
        account.change_pin(1234, 5678)
        self.assertTrue(account.verify_pin(5678))

    def test_change_pin_incorrect_old_pin(self):
        account = Account("John", "Doe", 1234, "1000000001")
        with self.assertRaises(ValueError) as context:
            account.change_pin(1111, 5678)
        self.assertEqual(str(context.exception), "Old PIN is incorrect")

    def test_change_pin_invalid_new_pin(self):
        account = Account("John", "Doe", 1234, "1000000001")
        with self.assertRaises(ValueError) as context:
            account.change_pin(1234, 56789)
        self.assertEqual(str(context.exception), "New PIN must be exactly 4 digits")

    def test_verify_pin_success(self):
        account = Account("John", "Doe", 1234, "1000000001")
        self.assertTrue(account.verify_pin(1234))

    def test_verify_pin_incorrect(self):
        account = Account("John", "Doe", 1234, "1000000001")
        self.assertFalse(account.verify_pin(1111))

if __name__ == "__main__":
    unittest.main()